scanner/decisions/0012-reindexing.md

```diff
@@ -0,0 +1,192 @@
+
+However, there are cons which outweigh the pros:
+
+* More data for Scanner V4 to track outside ClairCore
+  * This is not a blocker, but it is worth mentioning
+* Handling preexisting Index Reports
+  * This is the challenging aspect, which would block this
+  * There are a few options:
+    * Write a migration
+      * This involves doing essentially the same thing `ManifestScanned` does to determine the "versioned scanners"
+      * This may not be able to be pure PostgeSQL, as the `etag` is generated in Go, so that is an extra challenge
+    * Populate the table upon Index Report fetching
+      * If the entry does not exist in the table (the case for all Index Reports created prior to this release), then create the Index Report and populate the table
+      * This essentially forces a re-indexing of all previously indexed images, whether it is required or not, which may be costly
+
+### Random Deletions
+
+When it comes to variables outside "versioned scanners" there is no currently known clean way to handle the re-indexing.
+
+Something like creating our own "versioned scanner" to track other types of changes is not a solution:
+
+* Updating the version of the custom "versioned scanner" will only affect that scanner, and not the others which would all be skipped upon re-indexing
+* It is not feasible to track every single change in ClairCore which may affect Index Reports
+
+Instead, we will randomly delete manifests from the ClairCore database. Deleting a manifest from ClairCore
+deletes all data related to it, including the Index Report.
+Doing this solves both the remaining aspects of the re-indexing problem and the database ever-growing problem,
+as unneeded Index Reports are permanently deleted and needed reports may just be regenerated.
+
+**Note**: this approach also solves the "versioned scanner" update scenario; however, we opt to implement the [`checkManifest`](#reuse-claircores-checkmanifest-step)
+step for faster turnaround upon "versioned scanner" updates.
+
+Of course, determining a timeframe for deletion is non-trivial. Deleting too quickly means Index Reports may be corrected
+faster, but too many (probably) unnecessary requests to the image registry and (probably) unnecessary CPU/memory/disk usage.
+Deleting too slowly means resource usage will be much less, but it may take (too) long to correct the Index Report.
+Similarly, we would need to ensure not all Index Reports are all re-indexed at the same time (hence random deletions)
+to ensure spikes are minimized.
+
+The deletions will be chosen from a range between **one week** and **one month**. We believe this is a sufficient timeframe
+which balances speed and resource consumption.
+
+This will all be tracked in a new table (see [Manifest Metadata Table](#manifest-metadata-table) below).
+
+The table will be queried periodically for passed timestamps. These rows will be deleted and the related manifests/Index Reports will be deleted from ClairCore's tables.
```

Comment:
What will be doing the querying / deleting? Some goroutine in indexer/matcher? (if so how plan to coordinate multiple replicas) A cron type job in k8s?

Reply:
I am guessing it will be similar to goroutine "updaters"?

Reply:
goroutine in the indexer: https://github.com/stackrox/stackrox/pull/12648/files#diff-cbbf3957c55e04a712ce4daeff6aa9536f7929130c07fc20c0b61d3ebc824ff9

I figure it's ok to keep that unspecified in the ADR

Reply:
nit: Curious why keep something that like unspecified in the ADR? Seems like an appropriate thing to be included as it affects the architecture. May even influence future ADRs/designs where folks look for 'how to coordinate a background task amongst multiple replicas'.

Reply:
I saw this as a low level implementation detail, but I can add it